File: Computer Graphics/Line/DDA/DDA.java
// Java Code for DDA line generation
public class Solution {
 
  // function for rounding off the pixels
  public static int round(float n) {
    if (n - (int) n < 0.5)
      return (int) n;
    return (int) (n + 1);
  }
 
  // Function for line generation
  public static void DDALine(int x0, int y0, int x1, int y1) {
 
    // Calculate dx and dy
    int dx = x1 - x0;
    int dy = y1 - y0;
 
    int step;
 
    // If dx > dy we will take step as dx
    // else we will take step as dy to draw the complete
    // line
    if (Math.abs(dx) > Math.abs(dy))
      step = Math.abs(dx);
    else
      step = Math.abs(dy);
 
    // Calculate x-increment and y-increment for each step
    float x_incr = (float) dx / step;
    float y_incr = (float) dy / step;
 
    // Take the initial points as x and y
    float x = x0;
    float y = y0;
 
    for (int i = 0; i < step; i++) {
 
      // putpixel(round(x), round(y), WHITE);
      System.out.println(round(x) + " " + round(y));
      x += x_incr;
      y += y_incr;
      // delay(10);
    }
  }
 
  // Driver code
  public static void main(String[] args) {
 
    int x0 = 200, y0 = 180, x1 = 180, y1 = 160;
 
    // Function call
    DDALine(x0, y0, x1, y1);
 
  }
}

File: Computer Graphics/Line/DDA/DDA.py
# Python program for DDA line generation

from matplotlib import pyplot as plt

# DDA Function for line generation


def DDA(x0, y0, x1, y1):

	# find absolute differences
	dx = abs(x0 - x1)
	dy = abs(y0 - y1)

	# find maximum difference
	steps = max(dx, dy)

	# calculate the increment in x and y
	xinc = dx/steps
	yinc = dy/steps

	# start with 1st point
	x = float(x0)
	y = float(y0)

	# make a list for coordinates
	x_coorinates = []
	y_coorinates = []

	for i in range(steps):
		# append the x,y coordinates in respective list
		x_coorinates.append(x)
		y_coorinates.append(y)

		# increment the values
		x = x + xinc
		y = y + yinc

	# plot the line with coordinates list
	plt.plot(x_coorinates, y_coorinates, marker="o",
			markersize=1, markerfacecolor="green")
	plt.show()


# Driver code
if __name__ == "__main__":

	# coordinates of 1st point
	x0, y0 = 20, 20

	# coordinates of 2nd point
	x1, y1 = 60, 50

	# Function call
	DDA(x0, y0, x1, y1)

	# This code is contributed by 111arpit1
